flytekit/core/type_engine.py

```diff
@@ -486,14 +474,15 @@ def to_literal(self, ctx: FlyteContext, python_val: T, python_type: Type[T], exp
                 f"{type(python_val)} is not of type @dataclass, only Dataclasses are supported for "
                 f"user defined datatypes in Flytekit"
             )
-        if not self.is_serializable_class(type(python_val)):
-            raise TypeTransformerFailedError(
-                f"Dataclass {python_type} should be decorated with @dataclass_json or inherit DataClassJSONMixin to be "
-                f"serialized correctly"
-            )
+
         self._serialize_flyte_type(python_val, python_type)
 
-        json_str = python_val.to_json()  # type: ignore
+        if hasattr(python_val, "to_json"):
+            json_str = python_val.to_json()
+        else:
+            if not self._encoder.get(python_type):
+                self._encoder[python_type] = JSONEncoder(python_type)
+            json_str = self._encoder[python_type].encode(python_val)
```

Comment:
Nit: Usually Pythonic code prefers to "Ask for forgiveness"

```suggestion
            try:
                encoder = self._encoder[python_type]
            except KeyError:
                encoder = JSONEncoder(python_type)
                self._encoder[python_type] = encoder
            json_str = encoder.encode(python_val)
```

REF: https://devblogs.microsoft.com/python/idiomatic-python-eafp-versus-lbyl/

Reply:
thank you for this nice advice, I will also take notes about it!